qiskit/ignis/verification/tomography/fitters/gateset_fitter.py

```diff
@@ -0,0 +1,662 @@
+        return optimization_results
+
+
+class GaugeOptimize():
+    def __init__(self,
+                 Gs: List[np.array],
+                 initial_Gs_E: List[np.array],
+                 Fs: List[np.array],
+                 rho: np.array
+                 ):
+        """Initialize gauge optimizer fitter with the ideal and expected
+            outcomes.
+        Args:
+            Gs: The ideal expected gate matrices
+            initial_Gs_E: The experimentally-obtained gate approximations.
+            Fs: The SPAM circuit matrices
+            rho: The system's initial value (usually the |0> qubits)
+
+        Additional information:
+            Gauge optimization aims to find a basis in which the tomography
+            results are as close as possible to the ideal (noiseless) results
+
+            Given a gateset specification (E, rho, G1,...,Gn) and any
+            invertible matrix B, the gateset specification
+            (E*B^-1, B*rho, B*G1*B^-1,...,B*Gn*B^-1)
+            is indistinguishable from it by the tomography results.
+
+            B is called the gauge matrix and the goal of gauge optimization
+            is finding the B for which the resulting gateset description
+            is optimal in some sense; we choose to minimize the norm
+            difference between the gates found by experiment
+            and the "expected" gates in the ideal (noiseless) case.
+        """
+        self.Gs = Gs
+        self.Fs = Fs
+        self.initial_Gs_E = initial_Gs_E
+        self.d = np.shape(Gs[0])[0]
+        self.n = len(Gs)
+        self.rho = rho
+
+    def x_to_Gs_E(self, x: np.array) -> List[np.array]:
+        """Converts the gauge to the gateset defined by it
+                Args:
+                    x: An array representation of the B matrix
+
+                Returns:
+                    The gateset obtained from B
+
+                Additional information:
+                    Given a vector representation of B, this functions
+                    produces the list [B*G1*B^-1,...,B*Gn*B^-1]
+                    of gates correpsonding to the gauge B
+
+            """
+        B = np.array(x).reshape((self.d, self.d))
+        try:
+            BB = np.linalg.inv(B)
+            return [BB @ G @ B for G in self.initial_Gs_E]
+        except np.linalg.LinAlgError:
+            return np.inf
+
+    def obj_fn(self, x: np.array) -> float:
+        """The norm-based score function for the gauge optimizer
+            Args:
+                x: An array representation of the B matrix
+
+            Returns:
+                The sum of norm differences between the ideal gateset
+                and the one corresponding to B
+        """
+        Gs_E = self.x_to_Gs_E(x)
+        return sum([np.linalg.norm(G - G_E)
+                    for (G, G_E)
+                    in zip(self.Gs, Gs_E)])
+
+    def optimize(self) -> List[np.array]:
+        """The main optimization method
+            Returns:
```

Comment:
And for the rest of the doc strings in this class by the looks of things (and maybe some other classes)